sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/batch/BulkExecutor.java

```diff
@@ -593,4 +739,54 @@ private void onFlush() {
             logger.error("Callback invocation 'onFlush' failed.", t);
         }
     }
+
+    private static String getItemOperationDiagnostics(CosmosItemOperation operation) {
+
+        if (operation == FlushBuffersItemOperation.singleton()) {
+            return "ItemOperation[Type: Flush]";
+        }
+
+        StringBuilder sb = new StringBuilder();
+        sb
+            .append("ItemOperation[Type: ")
+            .append(operation.getOperationType().toString())
+            .append(", PK: ")
+            .append(operation.getPartitionKeyValue() != null ? operation.getPartitionKeyValue().toString() : "n/a")
+            .append(", id: ")
+            .append(operation.getId())
+            .append("]");
+
+        return sb.toString();
+    }
+
+    private static String getThreadInfo() {
+        StringBuilder sb = new StringBuilder();
+        Thread t = Thread.currentThread();
+        sb
+            .append("Thread[")
+            .append("Name: ")
+            .append(t.getName())
+            .append(",Group: ")
+            .append(t.getThreadGroup() != null ? t.getThreadGroup().getName() : "n/a")
+            .append(", isDaemon: ")
+            .append(t.isDaemon())
+            .append(", Id: ")
+            .append(t.getId())
+            .append("]");
+
+        return sb.toString();
+    }
+
+    private class SerializedEmitFailureHandler implements Sinks.EmitFailureHandler {
+
+        @Override
+        public boolean onEmitFailure(SignalType signalType, Sinks.EmitResult emitResult) {
+            logger.debug("SerializedEmitFailureHandler.onEmitFailure - Signal:{}, Result: {}", signalType, emitResult);
+            if (emitResult.equals(Sinks.EmitResult.FAIL_NON_SERIALIZED)) {
+                return true;
+            }
+
+            return false;
```

Comment:
maybe add error level log if can not retry.

Reply:
Done